Tabby: bump 0.7.0 -> 0.8.3 and add systemd service

ghthor opened the pull request:
## Description of changes

1. followup to #291744 
1. Upgraded tabby to release 0.8.3
2. Replaced static linking of vendored llama-cpp with dynamic link to llama-cpp from nixpkgs
3. Add systemd service for [tabby, tabby-scheduler] with timer configuration for tabby scheduler

### Upgraded to upstream 0.8.3 release

Added support for rocm AMD acceleration.

- [x] tested rocm acceleration locally with AMD Radeon RX 560 Series
- [x] tested cuda acceleration locally with NVIDIA Geforce RTX 2080 SUPER
- [x] tested darwin metal on an M1 macbook

Both tests used the new systemd service to run tabby

### Dynamic Linking w/ llama-cpp

I was having compile errors with the vendored copy of llama-cpp that tabbyml was statically compiling, so I switched to dynamic linking with the llama-cpp that is within nixpkgs.

### Systemd service

Example of the configuration required here.

https://github.com/ghthor/shrc/blob/18109599c526e580e54b11a16b6c1ef017777918/nix/cryptnix/configuration.nix#L434

```nix
  services.tabby = {
    enable = true;
    acceleration = "cuda";
    model = "TabbyML/StarCoder-3B";

    settings = {
      repositories = [
        {
          # Has to be created/updated manually, and permissions corrected
          git_url = "file:///var/lib/tabby/source_repo";
        }
      ];
    };
  };
```

Using the service makes it easy to enable a specific type of acceleration.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

ghthor commented:
@sersorrel I was able to test the rocm support on my local workstation, still pinging you if you'd like to test as well.

ghthor commented:
@happysalada ready for review

happysalada commented on pkgs/by-name/ta/tabby/package.nix:
```diff
@@ -1,36 +1,76 @@
+
+, llama-cpp
+# one of `[ null "rocm" "cuda" ]`
+, acceleration ? null
```
I would either
- not have that configurable and just use config.cudaSupport and rocmsupport
- or at least have the defaults being picked up by the config.

ghthor replied:
See comment in other thread

https://github.com/NixOS/nixpkgs/pull/292873#discussion_r1510808797

happysalada reviewed (commented):
This is looking good, thank you !
I have a couple of questions / comments

keysmashes reviewed (commented):
Looks good! I built with `nix-build -E 'with import ./. {}; tabby.override { acceleration = "rocm"; }'`, and `./result/bin/tabby serve --model TabbyML/StarCoder-1B --device rocm` appeared to work fine.

What's the purpose of the `modelDownload` config? I didn't explicitly run `tabby download`, it just downloaded it automatically.

ghthor commented on nixos/modules/services/misc/tabby.nix:
```diff
@@ -0,0 +1,183 @@
+          > TABBY_ROOT=/var/lib/tabby tabby download --model TabbyML/StarCoder-1B
+        '';
+      };
+      acceleration = lib.mkOption {
```
Yep, I think that makes sense, will fix this

ghthor replied:
I added some defaulting of the acceleration type in the presence of either `rocmSupport or cudaSupport` on `nixpkgs.config`. Even with that defaulting I have left `acceleration` as a service and package configuration value to handle the case when both of the above are present, as we can't make a clear decision and we do have to pick a single option. See this comment I've added to the code where the `llama-cpp` package is overriden.

https://github.com/NixOS/nixpkgs/pull/292873/commits/274d9b9ed96d4b32fa5173827814daaac914fa6e#diff-abdbe2b562428d2ed485d33ee478cf9e126faae74d5cc60b48284c7c254ddf33R86

ghthor commented on nixos/modules/services/misc/tabby.nix:
```diff
@@ -0,0 +1,183 @@
+
+      serviceEnv = lib.mkMerge [
+        {
+          TABBY_ROOT = "%S/tabby";
```
![Screenshot_2024-03-03_22-03-48](https://github.com/NixOS/nixpkgs/assets/160298/ac9072af-16bb-478c-a6db-91403e5ff78a)

https://www.freedesktop.org/software/systemd/man/latest/systemd.unit.html

I think `%S` just resolves to `/var/lib`.

Used this as reference https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/misc/ollama.nix#L48

And confirmed that `TABBY_ROOT=/var/lib/tabby` after systemd fills in `%S` with the following command

```
❯ sudo cat /proc/$(systemctl show --property MainPID tabby | cut -d= -f2)/environ | xargs --null --max-args=1 echo | grep ^TABBY
TABBY_DISABLE_USAGE_COLLECTION=1
TABBY_ROOT=/var/lib/tabby
```

ghthor commented on nixos/modules/services/misc/tabby.nix:
```diff
@@ -0,0 +1,183 @@
+
+      port = lib.mkOption {
+        type = types.port;
+        default = 11434;
```
It is, I didn't change it from the copy I made of ollama that I started from. I'll pick a different port

ghthor commented:
> What's the purpose of the modelDownload config? I didn't explicitly run tabby download, it just downloaded it automatically.

Yep, I can remove this extra configuration. I started using tabby before June 1, 2023 when they added this commit[1], didn't realize

[1] https://github.com/TabbyML/tabby/commit/ca077a34039a9e7bfe0de036802772d673b4b4e7

ghthor commented on nixos/modules/services/misc/tabby.nix:
```diff
@@ -0,0 +1,183 @@
+        StateDirectory = [ "tabby" ];
+        ConfigurationDirectory = [ "tabby" ];
+        DynamicUser = true;
+        User = "tabby";
```
https://github.com/NixOS/nixpkgs/pull/292873/commits/7cc929f28a2f1d162b92ea6ad0777ab37bd88f25

happysalada commented:
This looks good to me, happy to leave this open 1 more day to give some more time to other people to take a look.

de11n commented on pkgs/by-name/ta/tabby/package.nix:
```diff
@@ -1,36 +1,141 @@
+    len = builtins.length configured;
+    result = if len == 0 then "cpu" else (builtins.head configured);
+  in
+    # lib.warnIf (len > 1) "multiple acceleration methods available: ${toString availableAccelerations}" "cpu"
```
Is this comment meant to still be here?

ghthor replied:
Nope, will remove

de11n commented on pkgs/by-name/ta/tabby/package.nix:
```diff
@@ -1,36 +1,141 @@
+      For example:
+        services.tabby.acceleration = "cuda";
+      Or:
+        services.tabby.acceleration = "rocm";
```
It feels backward to mention NixOS module configuration in the warning of a specific package. Perhaps this validation deserves to be in the module instead. A shorter, less-verbose warning could be here instead.

ghthor replied:
Yeah that makes sense, I'll move this around

de11n replied:
Looks better. Thanks. It looks like you have a trailing whitespace somewhere, causing the CI to fail.

de11n reviewed (commented):
Extremely well done! A few small comments.

ghthor commented on nixos/modules/services/misc/tabby.nix:
```diff
@@ -0,0 +1,203 @@
+          `/var/lib/tabby` directory. You can then copy over the model manually
+          into `/var/lib/tabby`, update the model option to the name you just
+          downloaded and copied over then `nixos-rebuild switch` to start using
+          it.
```
Yeah, I had a hard time writing that out and it isn't a simple nor straighforward process to explain due to the implementations details of `DynamicUser` systemd config leaking. I'd be happy to revisit this and make it simpler to load models that the user already downloaded outside of the scope of the systemd service in a followup.

de11n replied:
I wouldn't consider it blocking. That's fine.

ghthor commented on nixos/modules/services/misc/tabby.nix:
```diff
@@ -0,0 +1,203 @@
+
+          - nixpkgs.config.cudaSupport
+          - nixpkgs.config.rocmSupport
+          - if stdenv.isDarwin && stdenv.isAarch64
```
It is supposed to convey where the package is auto configuring acceleration from in an attempt to help the user understand that in the face of multiple of these being set we're not able to make a good decision about which acceleration mode to use, thus suggest the user explicitly setting the acceleration method.

happysalada commented:
pkgs/by-name/ta/tabby/package.nix:
	70: Trailing whitespace

happysalada commented:
I just noticed the darwin x86_64 build is broken. We should at least mark it broken for that.
If you dont know about aarch64 i would mark it broken as well. If youve tested on darwin aarch64 then no problem leaving it as is.

ghthor commented:
Alright, little searching I found this https://github.com/NixOS/nixpkgs/pull/140325 for reference on how to mark the package as broken which I think points to it being a boolean flag in the meta object.

I don't have hardware to test linux-aarch64, but I will soon as that's what I'd like to try and run this on at work.

ghthor committed:
tabby: 0.7.0 -> 0.8.3

- Added support for AMD ROCm accleration
- Added support for Darwin Aarch64 Metal acceleration

https://github.com/TabbyML/tabby/releases/tag/v0.8.3

ghthor committed:
nixos/tabby: init module

- Enable tabby to run as a systemd service
- Document standard tabby configuration

ghthor commented:
Fixed commits to follow https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions

happysalada commented:
Perfect thanks ! I plan on merging this when i go home today.
